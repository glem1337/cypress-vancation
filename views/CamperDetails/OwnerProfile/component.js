import { Avatar, Button, Col, Row } from 'antd';
import PropTypes from 'prop-types';
import { FormattedMessage } from 'react-intl';

import { MAX_DESCRIPTION_VISIBLE_SYMBOLS } from 'constants/camperDetails/owner';

import userInitials from 'utils/userInitials';
import isPresent from 'utils/isPresent';

import SkeletonText from 'views/shared/SkeletonText';
import BtnGradient from 'views/shared/BtnGradient';
import Rating from './Rating';

const OwnerProfile = ({
  initialized,
  onRef,
  isCamperExist,
  isLoading,
  toggleDescription,
  description,
  allDescriptionVisible,
  onAskQuestion,
  owner,
  fullName,
  ratingProps,
}) => {
  if (!isCamperExist || isLoading) {
    return (
      <div ref={onRef} className="mb-40 mb-md-60">
        <Row gutter={24} align="bottom">
          <Col md={16} xl={10} className="mb-24 mb-md-0">
            <div className="skeleton__title w-100" />
            <SkeletonText />
          </Col>
          <Col md={8} xl={{ span: 6, offset: 4 }}>
            <div className="skeleton__title w-100" />
          </Col>
        </Row>
      </div>
    );
  }

  return (
    <div ref={onRef} className="mb-40 mb-md-60">
      {initialized && (
        <Row gutter={24} align="bottom">
          <Col md={16} xl={10} className="mb-24 mb-md-0">
            <div className="d-flex mb-16">
              <Avatar
                className="mr-16 flex-shrink-0"
                src={owner.avatarUrl}
                size={64}
                alt=""
              >
                {userInitials(...fullName)}
              </Avatar>
              <div>
                <div className="d-flex flex-wrap mb-8">
                  <p className="mr-4 text-headline font-700">
                    <FormattedMessage id="shared.ownedBy" />
                  </p>
                  <p className="text-headline in-blue-1000 font-700">
                    {owner.businessTitle || fullName.join(' ')}
                  </p>
                </div>
                <p className="">{owner.createdAt}</p>
              </div>
            </div>
            <div className="d-flex align-items-center flex-wrap mb-16">
              {isPresent(owner.averageRating) && (
                <Rating {...ratingProps} rating={owner.averageRating} />
              )}
              <p className="mr-16 text-subheader font-400 text-lowercase">
                <i className="icon icon-camper-f mr-8" />
                {owner.campersCount}
                {' '}
                <FormattedMessage id="shared.campers" />
              </p>
              {owner.idVerified && (
                <p className="mt-16 mt-md-0 text-subheader font-400">
                  <i className="icon icon-activate-f mr-8 in-azure-1000" />
                  <FormattedMessage id="shared.idVerified" />
                </p>
              )}
            </div>
            {isPresent(description) && (
              <p className="text-color-gray text-subheader font-400">
                {description}
              </p>
            )}
            {owner.description?.length > MAX_DESCRIPTION_VISIBLE_SYMBOLS && (
              <Button
                onClick={toggleDescription}
                className="mt-4 p-0 text-subheader font-400 h-auto"
                type="simple-text"
              >
                <FormattedMessage
                  id={allDescriptionVisible ? 'shared.hide' : 'shared.readMore'}
                />
              </Button>
            )}
          </Col>
          <Col md={8} xl={{ span: 6, offset: 4 }}>
            <BtnGradient
              onClick={onAskQuestion}
              size="large"
              className="min-w-180 main-btn--sm-100"
              text={{
                id: 'shared.askQuestion',
              }}
            />
          </Col>
        </Row>
      )}
    </div>
  );
};

OwnerProfile.defaultProps = {
  onRef: undefined,
  isLoading: false,
};

OwnerProfile.propTypes = {
  onRef: PropTypes.func,
  isLoading: PropTypes.bool,
  owner: PropTypes.shape().isRequired,
  initialized: PropTypes.bool.isRequired,
  isCamperExist: PropTypes.bool.isRequired,
  toggleDescription: PropTypes.func.isRequired,
  description: PropTypes.string.isRequired,
  fullName: PropTypes.arrayOf(PropTypes.string).isRequired,
  allDescriptionVisible: PropTypes.bool.isRequired,
  onAskQuestion: PropTypes.func.isRequired,
  ratingProps: PropTypes.shape().isRequired,
};

export default OwnerProfile;
