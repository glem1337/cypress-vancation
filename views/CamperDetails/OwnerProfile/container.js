import React from 'react';
import PropTypes from 'prop-types';
import * as R from 'ramda';
import { connect } from 'react-redux';

import ROUTES from 'constants/routes';
import {
  MAX_DESCRIPTION_VISIBLE_SYMBOLS,
  RATING_BOUNDARIES,
} from 'constants/camperDetails/owner';

import withIntersectionObserver from 'utils/hocs/withIntersectionObserver';
import redirect from 'utils/redirect';

import { showModal as showModalAction } from 'state/modal/actions';

import { loadingSelector } from 'state/data/selectors';

import { isUserLoggedInSelector } from 'state/concepts/session/selectors';

import { fetchCamperOwner } from 'state/concepts/camper/actions';
import { fetchCamperOwnerEndpoint } from 'state/concepts/camper/endpoints';
import {
  camperSelector,
  isCamperExistSelector,
} from 'state/concepts/camper/selectors';

import OwnerProfileComponent from './component';
import getUserFullName from '../../../utils/getUserFullName';

class OwnerProfile extends React.Component {
  static defaultProps = {
    isVisible: undefined,
    camper: null,
  };

  static propTypes = {
    isVisible: PropTypes.bool,
    fetchOwner: PropTypes.func.isRequired,
    showModal: PropTypes.func.isRequired,
    camperId: PropTypes.string.isRequired,
    isCamperExist: PropTypes.bool.isRequired,
    isUserLoggedIn: PropTypes.bool.isRequired,
    camper: PropTypes.shape(),
  };

  constructor(props) {
    super(props);

    this.state = {
      initialized: false,
      allDescriptionVisible: false,
    };
  }

  componentDidUpdate() {
    const { isVisible, fetchOwner, camperId, isCamperExist } = this.props;
    const { initialized } = this.state;

    if (isVisible && !initialized && isCamperExist) {
      fetchOwner(camperId);
      // eslint-disable-next-line react/no-did-update-set-state
      this.setState({
        initialized: true,
      });
    }
  }

  toggleDescription = () => {
    this.setState((prevState) => ({
      allDescriptionVisible: !prevState.allDescriptionVisible,
    }));
  };

  get description() {
    const { camper } = this.props;
    const { allDescriptionVisible } = this.state;

    const description = R.pathOr('', ['owner', 'description'], camper);

    if (
      allDescriptionVisible
      || description.length <= MAX_DESCRIPTION_VISIBLE_SYMBOLS
    ) {
      return description;
    }

    return `${description.substring(0, MAX_DESCRIPTION_VISIBLE_SYMBOLS)}...`;
  }

  onAskQuestionClickHandler = () => {
    const { isUserLoggedIn, showModal, camperId } = this.props;
    const { businessTitle } = this.owner;

    const placeholder = businessTitle || this.fullName.join(' ');

    if (isUserLoggedIn) {
      showModal({
        modalType: 'ASK_QUESTION_MODAL',
        modalProps: {
          camperId,
          placeholder,
        },
      });

      return;
    }

    redirect(ROUTES.LOGIN.PATH);
  };

  get owner() {
    const { camper } = this.props;

    return R.pathOr({}, ['owner'], camper);
  }

  get fullName() {
    return getUserFullName(['account', 'user'], this.owner);
  }

  get ratingProps() {
    const { averageRating } = this.owner;

    return {
      isHigh: averageRating >= RATING_BOUNDARIES.HIGH,
      isNormal:
        averageRating >= RATING_BOUNDARIES.NORMAL.START
        && averageRating <= RATING_BOUNDARIES.NORMAL.END,
      isMiddle:
        averageRating >= RATING_BOUNDARIES.MIDDLE.START
        && averageRating <= RATING_BOUNDARIES.MIDDLE.END,
      isLow: averageRating <= RATING_BOUNDARIES.LOW,
    };
  }

  render() {
    return (
      <OwnerProfileComponent
        {...this.state}
        {...this.props}
        toggleDescription={this.toggleDescription}
        description={this.description}
        onAskQuestion={this.onAskQuestionClickHandler}
        owner={this.owner}
        fullName={this.fullName}
        ratingProps={this.ratingProps}
      />
    );
  }
}

const mapStateToProps = (state, ownProps) => ({
  isUserLoggedIn: isUserLoggedInSelector(state),
  camper: camperSelector(state, ownProps?.camperId),
  isCamperExist: isCamperExistSelector(state, ownProps?.camperId),
  isLoading: loadingSelector(
    state,
    fetchCamperOwnerEndpoint(ownProps?.camperId).endpoint,
  ),
});

const mapDispatchToProps = {
  fetchOwner: fetchCamperOwner,
  showModal: showModalAction,
};

export { OwnerProfile as OwnerProfileContainer };
export default R.compose(
  connect(mapStateToProps, mapDispatchToProps),
  withIntersectionObserver(
    {
      isVisible: 0.0,
    },
    { rootMargin: '70% 0px' },
  ),
)(OwnerProfile);
