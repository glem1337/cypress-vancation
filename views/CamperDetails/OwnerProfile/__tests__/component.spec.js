import { shallow } from 'enzyme';

import OwnerProfile from '../component';

describe('OwnerProfile component tests', () => {
  const props = {
    initialized: true,
    onRef: jest.fn(),
    isCamperExist: true,
    isLoading: false,
    toggleDescription: jest.fn(),
    description: 'description',
    allDescriptionVisible: true,
    onAskQuestion: jest.fn(),
    owner: {
      avatarUrl: '/path/to/some/avatar/url',
      businessTitle: 'businessTitle',
      createdAt: 'createdAt',
      averageRating: 40,
      campersCount: 5,
      idVerified: true,
      description: 'description',
    },
    fullName: ['firstName', 'lastName'],
    ratingProps: {
      isHigh: false,
      isNormal: true,
      isMiddle: false,
      isLow: false,
    },
  };

  let component = null;

  beforeEach(() => {
    component = shallow(<OwnerProfile {...props} />);
  });

  describe('matches snapshot', () => {
    it('default', () => {
      expect(component).toMatchSnapshot();
    });

    it('when camper doesnt exist', () => {
      component.setProps({ isCamperExist: false });

      expect(component).toMatchSnapshot();
    });
  });
});
