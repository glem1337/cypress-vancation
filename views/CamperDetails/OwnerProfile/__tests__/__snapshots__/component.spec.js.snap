// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OwnerProfile component tests matches snapshot default 1`] = `
<div
  className="mb-40 mb-md-60"
>
  <Row
    align="bottom"
    gutter={24}
  >
    <Col
      className="mb-24 mb-md-0"
      md={16}
      xl={10}
    >
      <div
        className="d-flex mb-16"
      >
        <Avatar
          alt=""
          className="mr-16 flex-shrink-0"
          shape="circle"
          size={64}
          src="/path/to/some/avatar/url"
        >
          FL
        </Avatar>
        <div>
          <div
            className="d-flex flex-wrap mb-8"
          >
            <p
              className="mr-4 text-headline font-700"
            >
              <Memo(FormattedMessage)
                id="shared.ownedBy"
              />
            </p>
            <p
              className="text-headline in-blue-1000 font-700"
            >
              businessTitle
            </p>
          </div>
          <p
            className=""
          >
            createdAt
          </p>
        </div>
      </div>
      <div
        className="d-flex align-items-center flex-wrap mb-16"
      >
        <Rating
          isHigh={false}
          isLow={false}
          isMiddle={false}
          isNormal={true}
          rating={40}
        />
        <p
          className="mr-16 text-subheader font-400 text-lowercase"
        >
          <i
            className="icon icon-camper-f mr-8"
          />
          5
           
          <Memo(FormattedMessage)
            id="shared.campers"
          />
        </p>
        <p
          className="mt-16 mt-md-0 text-subheader font-400"
        >
          <i
            className="icon icon-activate-f mr-8 in-azure-1000"
          />
          <Memo(FormattedMessage)
            id="shared.idVerified"
          />
        </p>
      </div>
      <p
        className="text-color-gray text-subheader font-400"
      >
        description
      </p>
    </Col>
    <Col
      md={8}
      xl={
        Object {
          "offset": 4,
          "span": 6,
        }
      }
    >
      <BtnGradient
        className="min-w-180 main-btn--sm-100"
        isLoading={false}
        onClick={[MockFunction]}
        size="large"
        text={
          Object {
            "id": "shared.askQuestion",
          }
        }
      />
    </Col>
  </Row>
</div>
`;

exports[`OwnerProfile component tests matches snapshot when camper doesnt exist 1`] = `
<div
  className="mb-40 mb-md-60"
>
  <Row
    align="bottom"
    gutter={24}
  >
    <Col
      className="mb-24 mb-md-0"
      md={16}
      xl={10}
    >
      <div
        className="skeleton__title w-100"
      />
      <SkeletonText
        rows={2}
      />
    </Col>
    <Col
      md={8}
      xl={
        Object {
          "offset": 4,
          "span": 6,
        }
      }
    >
      <div
        className="skeleton__title w-100"
      />
    </Col>
  </Row>
</div>
`;
